packages/slice-machine/src/features/customTypes/customTypesBuilder/CustomTypeProvider.tsx

```diff
@@ -0,0 +1,96 @@
+import {
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import { useIsFirstRender } from "@prismicio/editor-support/React";
+import { CustomType } from "@prismicio/types-internal/lib/customtypes";
+
+import useSliceMachineActions from "@src/modules/useSliceMachineActions";
+import {
+  AutoSaveStatus,
+  useAutoSave,
+} from "@src/features/autoSave/useAutoSave";
+import { getFormat } from "@src/domain/customType";
+import { updateCustomType } from "@src/apiClient";
+
+import { CUSTOM_TYPES_MESSAGES } from "../customTypesMessages";
+
+type CustomTypeContext = {
+  customType: CustomType;
+  autoSaveStatus: AutoSaveStatus;
+  setCustomType: Dispatch<SetStateAction<CustomType>>;
+};
+
+type CustomTypeProviderProps = {
+  children: ReactNode | ((value: CustomTypeContext) => ReactNode);
+  initialCustomType: CustomType;
+};
+
+const CustomTypeContextValue = createContext<CustomTypeContext | undefined>(
+  undefined,
+);
+
+export function CustomTypeProvider(props: CustomTypeProviderProps) {
+  const { children, initialCustomType } = props;
+
+  const isFirstRender = useIsFirstRender();
+  const [customType, setCustomType] = useState(initialCustomType);
+  const format = getFormat(customType);
+  const customTypeMessages = CUSTOM_TYPES_MESSAGES[format];
+  const { autoSaveStatus, setNextSave } = useAutoSave({
+    errorMessage: customTypeMessages.autoSaveFailed,
+  });
+  const { saveCustomTypeSuccess } = useSliceMachineActions();
+
+  useEffect(
+    () => {
+      // Prevent a save to be triggered on first render
+      if (!isFirstRender) {
+        setNextSave(async () => {
+          const { errors } = await updateCustomType(customType);
+
+          if (errors.length > 0) {
+            throw errors;
+          }
+
+          // Update available custom types store with new custom type
+          saveCustomTypeSuccess(customType);
+        });
+      }
+    },
+    // Prevent saveCustomTypeSuccess from triggering an infinite loop
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [customType, setNextSave],
```

Comment:
Is it better to remove it from the deps or the wrap it with a useCallback?

Reply:
I could try to use `useStableCallback` from the editor support, good idea!

Reply:
It's working with `useStableCallback`:

<img width="617" alt="Screenshot 2024-01-17 at 14 55 57" src="https://github.com/prismicio/slice-machine/assets/19946868/43159018-9e16-4c69-b4f4-271608a02ed9">

I did it in the second PR with both providers to freeze this one, since the review is done now.